baybe/symmetries.py

```diff
@@ -0,0 +1,439 @@
+class Symmetry(SerialMixin, ABC):
+    """Abstract base class for symmetries.
+
+    Symmetry is a concept that can be used to configure the modelling process in the
+    presence of invariances.
+    """
+
+    use_data_augmentation: bool = field(
+        default=True, validator=instance_of(bool), kw_only=True
+    )
+    """Flag indicating whether data augmentation would be used with surrogates that
+    support this."""
+
+    @property
+    @abstractmethod
+    def parameter_names(self) -> tuple[str, ...]:
+        """The names of the parameters affected by the symmetry."""
+
+    def summary(self) -> dict:
+        """Return a custom summarization of the symmetry."""
+        symmetry_dict = dict(
+            Type=self.__class__.__name__, Affected_Parameters=self.parameter_names
+        )
+        return symmetry_dict
+
+    @abstractmethod
+    def augment_measurements(
+        self, df: pd.DataFrame, parameters: Iterable[Parameter]
+    ) -> pd.DataFrame:
+        """Augment the given measurements according to the symmetry.
+
+        Args:
+            df: The dataframe containing the measurements to be augmented.
+            parameters: Parameter objects carrying additional information (might not be
+                needed by all augmentation implementations).
+
+        Returns:
+            The augmented dataframe including the original measurements.
+        """
+
+    def validate_searchspace_context(self, searchspace: SearchSpace) -> None:
+        """Validate that the symmetry is compatible with the given searchspace.
+
+        Args:
+            searchspace: The searchspace to validate against.
+
+        Raises:
+            ValueError: If the symmetry affects parameters not present in the
+                searchspace.
+        """
+        parameters_missing = set(self.parameter_names).difference(
+            searchspace.parameter_names
+        )
+        if parameters_missing:
+            raise ValueError(
+                f"The symmetry of type {self.__class__.__name__} was set up with at "
+                f"least one parameter which is not present in the searchspace: "
+                f"{parameters_missing}."
+            )
+
+
+@define(frozen=True)
+class PermutationSymmetry(Symmetry):
+    """Class for representing permutation symmetries.
+
+    A permutation symmetry expresses that certain parameters can be permuted without
+    affecting the outcome of the model. For instance, this is the case if
+    $f(x,y) = f(y,x)$.
+    """
+
+    _parameter_names: tuple[str, ...] = field(
+        alias="parameter_names",
+        converter=Converter(normalize_str_sequence, takes_self=True, takes_field=True),  # type: ignore
+        validator=(  # type: ignore
+            validate_unique_values,
+            deep_iterable(
+                member_validator=instance_of(str), iterable_validator=min_len(2)
+            ),
+        ),
+    )
+    """The names of the parameters affected by the symmetry."""
+
+    @override
```

Comment:
The definition order is a bit messed up here, because the property is in between attributes.

Let's stick with the following pattern (where initialization-related stuff comes first)
* attributes
* validators / post-init
* properties
* methods